cpp/include/cudf/join/hash_join.hpp

```diff
@@ -0,0 +1,325 @@
+            rmm::device_async_resource_ref mr      = cudf::get_current_device_resource_ref()) const;
+
+  /**
+   * Returns the exact number of matches (rows) when performing an inner join with the specified
+   * probe table.
+   *
+   * @param probe The probe table, from which the tuples are probed
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   *
+   * @throw cudf::logic_error If the input probe table has nulls while this hash_join object was not
+   * constructed with null check.
+   *
+   * @return The exact number of output when performing an inner join between two tables with
+   * `build` and `probe` as the join keys .
+   */
+  [[nodiscard]] std::size_t inner_join_size(
+    cudf::table_view const& probe, rmm::cuda_stream_view stream = cudf::get_default_stream()) const;
+
+  /**
+   * Returns the exact number of matches (rows) when performing a left join with the specified probe
+   * table.
+   *
+   * @param probe The probe table, from which the tuples are probed
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   *
+   * @throw cudf::logic_error If the input probe table has nulls while this hash_join object was not
+   * constructed with null check.
+   *
+   * @return The exact number of output when performing a left join between two tables with `build`
+   * and `probe` as the join keys .
+   */
+  [[nodiscard]] std::size_t left_join_size(
+    cudf::table_view const& probe, rmm::cuda_stream_view stream = cudf::get_default_stream()) const;
+
+  /**
+   * Returns the exact number of matches (rows) when performing a full join with the specified probe
+   * table.
+   *
+   * @param probe The probe table, from which the tuples are probed
+   * @param stream CUDA stream used for device memory operations and kernel launches
+   * @param mr Device memory resource used to allocate the intermediate table and columns' device
+   * memory.
+   *
+   * @throw cudf::logic_error If the input probe table has nulls while this hash_join object was not
+   * constructed with null check.
+   *
+   * @return The exact number of output when performing a full join between two tables with `build`
+   * and `probe` as the join keys .
+   */
+  [[nodiscard]] std::size_t full_join_size(
+    cudf::table_view const& probe,
+    rmm::cuda_stream_view stream      = cudf::get_default_stream(),
+    rmm::device_async_resource_ref mr = cudf::get_current_device_resource_ref()) const;
+
+ private:
+  std::unique_ptr<impl_type const> _impl;
+};
+
+/**
+ * @brief Distinct hash join that builds hash table in creation and probes results in subsequent
+ * `*_join` member functions
+ *
+ * This class enables the distinct hash join scheme that builds hash table once, and probes as many
+ * times as needed (possibly in parallel).
+ *
+ * @note Behavior is undefined if the build table contains duplicates.
+ * @note All NaNs are considered as equal
+ */
+class distinct_hash_join {
```

Comment:
let's move distinct_hash_join to its dedicated header as well since the data structure used is different from the normal `hash_join`